src/AspNetCoreModuleV2/CommonLib/hostfxr_utility.cpp

```diff
@@ -3,344 +3,197 @@
         //
-        if (is_regular_file(processPath))
+        if (is_regular_file(executablePath))
         {
-            RETURN_IF_FAILED(GetStandaloneHostfxrParameters(
-                processPath.c_str(),
-                pcwzApplicationPhysicalPath,
-                arguments.c_str(),
-                pStruHostFxrDllLocation,
-                pdwArgCount,
-                pbstrArgv));
-
-            RETURN_IF_FAILED(pStruExeAbsolutePath->Copy(processPath.c_str()));
+            auto applicationDllPath = executablePath;
+            applicationDllPath.replace_extension(".dll");
+
+            LOG_INFOF("Checking application.dll at %S", applicationDllPath.c_str());
+            if (!is_regular_file(applicationDllPath))
+            {
+                throw StartupParametersResolutionException(format(L"Application .dll was not found at %s", applicationDllPath.c_str()));
+            }
+
+            hostFxrDllPath = executablePath.parent_path() / "hostfxr.dll";
+            LOG_INFOF("Checking hostfxr.dll at %S", hostFxrDllPath.c_str());
+            if (is_regular_file(hostFxrDllPath))
+            {
+                LOG_INFOF("hostfxr.dll found app local at '%S', treating application as standalone", hostFxrDllPath.c_str());
+            }
+            else
+            {
+                LOG_INFOF("hostfxr.dll found app local at '%S', treating application as portable with launcher", hostFxrDllPath.c_str());
+
+                // passing "dotnet" here because we don't know where dotnet.exe should come from
+                // so trying all fallbacks is appropriate
+                if (dotnetExePath.empty())
+                {
+                    dotnetExePath = GetAbsolutePathToDotnet(applicationPhysicalPath, L"dotnet");
+                }
+                executablePath = dotnetExePath;
+                hostFxrDllPath = GetAbsolutePathToHostFxr(dotnetExePath);
+            }
+
+            ParseHostfxrArguments(
+                applicationDllPath.generic_wstring() + L" " + expandedApplicationArguments,
+                executablePath,
+                applicationPhysicalPath,
+                arguments);
         }
         else
         {
             //
             // If the processPath file does not exist and it doesn't include dotnet.exe or dotnet
             // then it is an invalid argument.
             //
-            hr = HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND);
-            EventLog::Error(
-                ASPNETCORE_EVENT_INVALID_PROCESS_PATH,
-                ASPNETCORE_EVENT_INVALID_PROCESS_PATH_MSG,
-                processPath.c_str(),
-                hr);
-            return hr;
+            throw StartupParametersResolutionException(format(L"Executable was not found at '%s'", executablePath.c_str()));
         }
     }
+}
 
-    return S_OK;
+BOOL
+HOSTFXR_UTILITY::IsDotnetExecutable(const std::filesystem::path & dotnetPath)
```

Comment:
Nit: Consider dropping this this function entirely and just calling ends_with. It's only used once.